### ISSUE TITLE
AdaBoost and RandomForest class weights

### ISSUE THREAD

**Issue by jknightmmcs** [mlpack/mlpack#2750]
Hey, people! Not sure if this is an issue, but I still don't understand the right way to do it.

Let's say we have imbalanced dataset (like one class being about 3-4x the other). Surely, training on such a dataset fails.

I've tried to fix that using _weights_ parameter for RandomForest (size equals a number of training samples, all weights equal **1** for class 0 and **num_0/num_1** for class 1). Didn't help as much, probabilities for success are totally the same.

Tried doing the same thing for DecisionTree weak learner for AdaBoost, didn't help at all. Then, after a thorough RTFM I have found an understanding that _weights_ parameter is ignored for AdaBoost weak learners.

So, question is: how to do imbalanced dataset learning with RandomForest and AdaBoost right? Is it even possible?

**Comment by rcurtin** [comment]
Hey @jknightmmcs, I think that this support should work right.  With `AdaBoost`, there is actually no option to pass weights along with labels---the AdaBoost algorithm reweights misclassified points at each iteration of boosting, so it should be able to achieve perfect accuracy on the training set, under the single exception where there are two training points that are exactly the same points but have different labels.  `RandomForest` does support weighted training, like you found out.  Are you sure you set the weights right and other hyperparameters too?  Does it give the same results as if you duplicated each of the points in class 1 3 or 4 times (that should be equivalent to weighting but of course it is not what we would actually like to do in practice)?  Consider using `Probabilities()` and not `Predict()` and tuning the threshold for classification for the weighted and unweighted random forests.

**Comment by jknightmmcs** [comment]
@rcurtin, thanks for the fast response! Will think on what you said =3
